src/features/ci-cd/hooks/service-installs.ts

```diff
@@ -155,27 +231,68 @@ hooks.addPureHook('POST', 'resin', 'device', {
 });
 
 hooks.addPureHook('PATCH', 'resin', 'device', {
-	POSTRUN: async (args) => {
-		const affectedIds = args.request.affectedIds!;
+	PRERUN: async (args) => {
+		const affectedDeviceIds = args.request.affectedIds!;
 
-		// We need to delete all service_install resources for the current device and
+		// We need to delete all service_install resources for the current device / app and
 		// create new ones for the new application (if the device is moving application)
-		if (
-			args.request.values.belongs_to__application != null &&
-			affectedIds.length !== 0
-		) {
-			await args.api.delete({
-				resource: 'service_install',
+		const newAppId = args.request.values.belongs_to__application;
+		if (newAppId && affectedDeviceIds.length !== 0) {
+			const createHandler = async (
+				api: sbvrUtils.PinejsClient,
+				deviceIds: number[],
+				serviceIds: number[],
+			) => {
+				await Promise.all(
+					deviceIds.map(async (deviceId) => {
+						await Promise.all(
+							serviceIds.map(async (serviceId) => {
+								// Create a service_install for this pair of service and device
+								await api.post({
+									resource: 'service_install',
+									body: {
+										device: deviceId,
+										installs__service: serviceId,
+									},
+									options: { returnResource: false },
+								});
+							}),
+						);
+					}),
+				);
+			};
+
+			// get devices that are currently on a different app and delete those service installs
+			const devices = (await args.api.get({
+				resource: 'device',
 				options: {
+					$select: ['id', 'belongs_to__application'],
 					$filter: {
-						device: { $in: affectedIds },
+						id: { $in: affectedDeviceIds },
 					},
 				},
-			});
-			await createAppServiceInstalls(
+			})) as Array<{
+				id: number;
+				belongs_to__application: { __id: number };
+			}>;
```

Comment:
It'd be better to fetch the app ids directly and remove the need to fetch the extra data and then process it locally:
```ts
			const apps = (await args.api.get({
				resource: 'application',
				options: {
					$select: ['id'],
					$filter: {
						owns__device: {
							$any: {
								$alias: 'd',
								$expr: { d: { id: { $in: affectedDeviceIds }, }
						}
					},
				},
			})) as Array<Pick<Application, 'id'>>;